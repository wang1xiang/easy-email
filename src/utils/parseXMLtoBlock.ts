import { BlocksMap } from '@/components/core/blocks';
import mjml from 'mjml-browser';
import { IBlockData } from '@/typings';
import { BlockType, BasicType } from './../constants';
import { MjmlToJson } from './MjmlToJson';

const domParser = new DOMParser();
export function parseXMLtoBlock(text: string) {
  const dom = domParser.parseFromString(text, 'text/xml');
  const root = dom.firstChild as Element;
  if (!(dom.firstChild instanceof Element)) {
    throw new Error('Invalid content');
  }
  if (root.tagName === 'mjml') {
    const { json } = mjml(text, {
      validationLevel: 'soft',
    });
    const parseValue = MjmlToJson(json);
    return parseValue;
  }

  const transform = (node: Element): IBlockData => {
    if (node.tagName === 'parsererror') {
      throw new Error('Invalid content');
    }
    const attributes: IBlockData['attributes'] = {};
    node.getAttributeNames().forEach((name) => {
      attributes[name] = node.getAttribute(name);
    });
    const type = node.tagName.replace('mj-', '');

    if (!BlocksMap.findBlockByType(type)) {
      if (!node.parentElement || node.parentElement.tagName !== 'mj-text')
        throw new Error('Invalid content');
    }

    const block: IBlockData = {
      type: type as BlockType,
      attributes: attributes,
      data: {
        value: {
          content: node.textContent?.trim(),
        },
      },
      children: [...node.children]
        .filter((item) => item instanceof Element)
        .map(transform as any),
    };

    switch (type) {
      case BasicType.TEXT:
        block.data.value.content = node.innerHTML;
        block.children = [];
    }

    return block;
  };

  return transform(root);
}
