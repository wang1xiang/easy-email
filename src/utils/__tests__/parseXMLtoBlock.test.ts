import { parseXMLtoBlock } from '../parseXMLtoBlock';

const mjmlText = `

<mjml>
<mj-head>

<mj-html-attributes>
<mj-html-attribute class="easy-email" multiple-attributes="false" attribute-name="text-color" text-color="#000000"></mj-html-attribute>
<mj-html-attribute class="easy-email" multiple-attributes="false" attribute-name="font-family" font-family="'Hiragino Kaku Gothic ProN','Hiragino Sans',Meiryo"></mj-html-attribute>
<mj-html-attribute class="easy-email" multiple-attributes="false" attribute-name="responsive" responsive="true"></mj-html-attribute>

</mj-html-attributes>



    <mj-breakpoint width="480px" />
  <mj-attributes>

    <mj-all font-family="'Hiragino Kaku Gothic ProN','Hiragino Sans',Meiryo" />
    <mj-text color="#000000" />


  </mj-attributes>
</mj-head>
<mj-body background-color="#efeeea" css-class="mjml-body">

    <mj-wrapper padding="20px 0px 20px 0px" border="none" direction="ltr" text-align="center" background-color="rgba(255, 255, 255, 0)">

    <mj-section padding="20px 0px 0px 0px" background-repeat="repeat" background-size="auto" background-position="top center" border="none" direction="ltr" text-align="center">

    <mj-column padding="0px 0px 0px 0px" border="none" vertical-align="top">

<mj-image align="center" height="auto" padding="10px 25px 10px 25px" width="150px" src="https://i.postimg.cc/63Bzt8fw/Tabechoku-Tastin-collab.png">

</mj-image>


<mj-text font-size="13px" padding="10px 25px 10px 25px" line-height="1" align="center">
 この度は、#cookforJapn by 食べチョクのご利用ありがとうございます。<div>シェフからのメッセージをお楽しみください。</div>
</mj-text>

    </mj-column>

    </mj-section>

    </mj-wrapper>


    <mj-section padding="20px 0px 20px 0px" background-repeat="repeat" background-size="auto" background-position="top center" border="none" direction="ltr" text-align="center" background-color="rgba(255, 255, 255, 1)">

    <mj-column padding="0px 0px 0px 0px" border="none" vertical-align="top">

<mj-text font-size="16px" padding="10px 25px 10px 25px" line-height="1" align="left">
 <b>飲み物にもこだわる大切さ</b>
</mj-text>


<mj-text font-size="13px" padding="0px 25px 0px 25px" line-height="1" align="left">
 担当料理: 日本酒
</mj-text>


<mj-image align="center" height="auto" padding="10px 25px 10px 25px" width="1000px" src="https://i.postimg.cc/brF57GLv/image.png" href="https://soundcloud.com/kyle-comstock-126175966/food-podcast">

</mj-image>


<mj-text font-size="13px" padding="10px 25px 10px 25px" line-height="1" align="left">
 日本酒と日本酒を育てた方や育った場所に寄り添い、新しいことも積極的に取り入れ実験し、挑戦なさる橋野さん。そんな橋野さんの味わい深くも、モダンなお考えや想い、ストーリーをお届けします。
</mj-text>


<mj-button align="center" background-color="rgba(50, 73, 103, 1)" color="#ffffff" font-size="13px" font-weight="normal" border-radius="3px" padding="10px 25px 10px 25px" inner-padding="10px 25px 10px 25px" line-height="120%" target="_blank" vertical-align="middle" border="none" text-align="center" href="#">
 🎧視聴する
</mj-button>

    </mj-column>

    </mj-section>

</mj-body>
</mjml>

`;
describe('Test parseXml', () => {
  const instance = parseXMLtoBlock(mjmlText);

  it('should render  as expected', () => {
    expect(instance).toMatchSnapshot();
  });

  it('should render  as expected', () => {
    expect(
      parseXMLtoBlock(`
    <mj-text font-size="13px" padding="0px 0px 0px 0px" line-height="1" align="left">
     <div style="color:#fff;">Make it easy for everyone to compose emails!</div>
    </mj-text>
  `)
    ).toMatchSnapshot();
  });

  it('should throw error when xml is unexpected', () => {
    expect(() => parseXMLtoBlock('<mj-button><mj-button>')).toThrowError(
      'Invalid content'
    );
    expect(() => parseXMLtoBlock('<html></html>')).toThrowError(
      'Invalid content'
    );
    expect(() =>
      parseXMLtoBlock(`
    <mj-button font-size="13px" padding="0px 0px 0px 0px" line-height="1" align="left">
     <div style="color:#fff;">Make it easy for everyone to compose emails!</div>
    </mj-button>
  `)
    ).toThrowError('Invalid content');
  });
});
