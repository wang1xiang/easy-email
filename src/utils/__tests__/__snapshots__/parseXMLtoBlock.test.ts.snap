// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test parseXml should render  as expected 1`] = `
Object {
  "attributes": Object {
    "background-color": "#efeeea",
    "css-class": "mjml-body",
    "width": "600px",
  },
  "children": Array [
    Object {
      "attributes": Object {
        "background-color": "rgba(255, 255, 255, 0)",
        "border": "none",
        "direction": "ltr",
        "padding": "20px 0px 20px 0px",
        "text-align": "center",
      },
      "children": Array [
        Object {
          "attributes": Object {
            "background-position": "top center",
            "background-repeat": "repeat",
            "background-size": "auto",
            "border": "none",
            "direction": "ltr",
            "padding": "20px 0px 0px 0px",
            "text-align": "center",
          },
          "children": Array [
            Object {
              "attributes": Object {
                "border": "none",
                "padding": "0px 0px 0px 0px",
                "vertical-align": "top",
              },
              "children": Array [
                Object {
                  "attributes": Object {
                    "align": "center",
                    "height": "auto",
                    "padding": "10px 25px 10px 25px",
                    "src": "https://i.postimg.cc/63Bzt8fw/Tabechoku-Tastin-collab.png",
                    "width": "150px",
                  },
                  "children": Array [],
                  "data": Object {
                    "value": Object {},
                  },
                  "type": "image",
                },
                Object {
                  "attributes": Object {
                    "align": "center",
                    "font-size": "13px",
                    "line-height": "1",
                    "padding": "10px 25px 10px 25px",
                  },
                  "children": Array [],
                  "data": Object {
                    "value": Object {
                      "content": "この度は、#cookforJapn by 食べチョクのご利用ありがとうございます。<div>シェフからのメッセージをお楽しみください。</div>",
                    },
                  },
                  "type": "text",
                },
              ],
              "data": Object {
                "value": Object {},
              },
              "type": "column",
            },
          ],
          "data": Object {
            "value": Object {
              "noWrap": false,
            },
          },
          "type": "section",
        },
      ],
      "data": Object {
        "value": Object {},
      },
      "type": "wrapper",
    },
    Object {
      "attributes": Object {
        "background-color": "rgba(255, 255, 255, 1)",
        "background-position": "top center",
        "background-repeat": "repeat",
        "background-size": "auto",
        "border": "none",
        "direction": "ltr",
        "padding": "20px 0px 20px 0px",
        "text-align": "center",
      },
      "children": Array [
        Object {
          "attributes": Object {
            "border": "none",
            "padding": "0px 0px 0px 0px",
            "vertical-align": "top",
          },
          "children": Array [
            Object {
              "attributes": Object {
                "align": "left",
                "font-size": "16px",
                "line-height": "1",
                "padding": "10px 25px 10px 25px",
              },
              "children": Array [],
              "data": Object {
                "value": Object {
                  "content": "<b>飲み物にもこだわる大切さ</b>",
                },
              },
              "type": "text",
            },
            Object {
              "attributes": Object {
                "align": "left",
                "font-size": "13px",
                "line-height": "1",
                "padding": "0px 25px 0px 25px",
              },
              "children": Array [],
              "data": Object {
                "value": Object {
                  "content": "担当料理: 日本酒",
                },
              },
              "type": "text",
            },
            Object {
              "attributes": Object {
                "align": "center",
                "height": "auto",
                "href": "https://soundcloud.com/kyle-comstock-126175966/food-podcast",
                "padding": "10px 25px 10px 25px",
                "src": "https://i.postimg.cc/brF57GLv/image.png",
                "width": "1000px",
              },
              "children": Array [],
              "data": Object {
                "value": Object {},
              },
              "type": "image",
            },
            Object {
              "attributes": Object {
                "align": "left",
                "font-size": "13px",
                "line-height": "1",
                "padding": "10px 25px 10px 25px",
              },
              "children": Array [],
              "data": Object {
                "value": Object {
                  "content": "日本酒と日本酒を育てた方や育った場所に寄り添い、新しいことも積極的に取り入れ実験し、挑戦なさる橋野さん。そんな橋野さんの味わい深くも、モダンなお考えや想い、ストーリーをお届けします。",
                },
              },
              "type": "text",
            },
            Object {
              "attributes": Object {
                "align": "center",
                "background-color": "rgba(50, 73, 103, 1)",
                "border": "none",
                "border-radius": "3px",
                "color": "#ffffff",
                "font-size": "13px",
                "font-weight": "normal",
                "href": "#",
                "line-height": "120%",
                "padding": "10px 25px 10px 25px",
                "target": "_blank",
                "text-align": "center",
                "vertical-align": "middle",
              },
              "children": Array [],
              "data": Object {
                "value": Object {
                  "content": "🎧視聴する",
                },
              },
              "type": "button",
            },
          ],
          "data": Object {
            "value": Object {},
          },
          "type": "column",
        },
      ],
      "data": Object {
        "value": Object {
          "noWrap": false,
        },
      },
      "type": "section",
    },
  ],
  "data": Object {
    "value": Object {
      "breakpoint": "480px",
      "font-family": "'Hiragino Kaku Gothic ProN','Hiragino Sans',Meiryo",
      "font-size": "14px",
      "fonts": Array [],
      "headAttributes": "",
      "headStyles": Array [],
      "line-height": "1.7",
      "responsive": true,
      "text-color": "#000000",
    },
  },
  "type": "page",
}
`;

exports[`Test parseXml should render  as expected 2`] = `
Object {
  "attributes": Object {
    "align": "left",
    "font-size": "13px",
    "line-height": "1",
    "padding": "0px 0px 0px 0px",
  },
  "children": Array [],
  "data": Object {
    "value": Object {
      "content": "
     <div style=\\"color:#fff;\\">Make it easy for everyone to compose emails!</div>
    ",
    },
  },
  "type": "text",
}
`;
